File: Simulator/Effects/Wave.cs
using System;
using System.Windows.Media;

namespace Simulator.Effects
{
    public class Wave : Effect
    {
        private float phase;

        public float WaveLength { get; set; } = 18f;
        public float WaveSpeed { get; set; } = 0.0004f;
        public Color LowColor { get; set; } = Color.FromRgb(127, 80, 0);
        public Color HighColor { get; set; } = Colors.Yellow;

        public Wave(Matrix matrix) : base(matrix)
        {
            Background = Colors.DarkGray;
        }

        protected override void LoopInternal(long dt)
        {
            phase = (phase + dt * WaveSpeed) % 1f;
            Draw();
        }

        private void Draw()
        {
            for (var x = 0; x < matrix.Width; x++)
            {
                var dist = Math.Abs(x - WaveLength * phase) % WaveLength;
                dist = Math.Min(dist, WaveLength - dist);
                var ratio = dist < 1f ? 1f : 1f / dist;
                ratio = Math.Clamp(ratio, 0f, 1f);
                var color = LowColor.Blend(HighColor, ratio);

                for (var y = 0; y < matrix.Height; y++)
                    if ((x + y) % 2 != 0)
                        matrix[x, y].Color = color;
            }
        }
    }
}

File: NyMatrix/Wave.cpp
#include "Wave.h"

void Wave::setupInternal()
{
}

void Wave::loopInternal(unsigned long dt)
{
    phase = fmod(phase + dt * waveSpeed, 1.0f);
    draw();
}

void Wave::draw()
{
    clear();

    for (int x = 0; x < MATRIX_WIDTH; x++)
    {
        auto dist = fmod(abs(x - waveLength * phase), waveLength);
        dist = min(dist, waveLength - dist);
        auto ratio = dist < 1.0f ? 1.0f : 1.0f / dist;
        ratio = max(0.0f, min(1.0f, ratio));
        auto color = RgbColor::LinearBlend(lowColor, highColor, ratio);

        for (int y = 0; y < MATRIX_HEIGHT; y++)
            if ((x + y) % 2 != 0)
                matrix->SetPixelColor(topo->Map(x, y), color);
    }

    dim();
}
